### PR TITLE
Network layer

### PR DESCRIPTION
**Pull request by JonathanUsername**
Add Network Layer with requests store for manipulating requests and adding listeners.

### CODE REVIEW

File: src/mutate.js [hunk 1]
```diff
@@ -48,18 +48,66 @@ export type MutationConfig = {
...
 export type Mutate = (config: MutationConfig) => Promise<*>;
 
+type PromiseParams = {env: Relay.Environment, config: MutationConfig} | Function;
 
+// TODO: Does this need a polyfill for node?
+class MutationPromise extends Promise {
+    abort: () => void = () => {};
+    onProgress: () => void = () => {};
+    env: Relay.Environment;
+    mutation: Relay.GraphQLMutation;
```
Comment (sdcooke) [inline src/mutate.js]: _env and _mutation - we shouldn't be exposing these externally

File: src/mutate.js [hunk 2]
```diff
@@ -48,18 +48,66 @@ export type MutationConfig = {
 
 export type Mutate = (config: MutationConfig) => Promise<*>;
 
+type PromiseParams = {env: Relay.Environment, config: MutationConfig} | Function;
 
+// TODO: Does this need a polyfill for node?
```
Comment (sdcooke) [inline src/mutate.js]: Remove TODO

File: src/mutate.js [hunk 3]
```diff
@@ -48,18 +48,66 @@ export type MutationConfig = {
...
+        return this; // To allow for method chaining
+    }
+
+    get request() {
+        try {
+            const mutationId = this.mutation._transaction.id;
+            const request = this.env._requests.get(mutationId);
+            return request;
+        } catch (e) {
+            console.error('No request or mutation');
```
Comment (sdcooke) [inline src/mutate.js]: Remove console log

File: src/networkLayer/client.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
+/* @flow */
+import Relay from 'react-relay/classic';
+import RequestObject from './requestObject';
+
+export type ClientNetworkLayerOpts = {
```
Comment (sdcooke) [inline src/networkLayer/client.js]: A bunch of this stuff shouldn't be in relay-helpers - it's very Mixcloud specific

File: src/networkLayer/requestObject.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+type RequestObjectOpts = {
+    id: string,
+    request: XMLHttpRequest,
+    removeSelf: () => void
+};
+
+class RequestObject {
+    id: string;
+    request: XMLHttpRequest;
```
Comment (sdcooke) [inline src/networkLayer/requestObject.js]: _xhr? This shouldn't expost the XMLHttpRequest

File: src/networkLayer/requestObject.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    constructor(opts: RequestObjectOpts) {
+        Object.assign(this, opts);
+    }
+
+    abort() {
+        this.request.abort();
+        this.removeSelf();
+    }
+
+    onProgress(cb: Function) {
```
Comment (sdcooke) [inline src/networkLayer/requestObject.js]: You can make a more specific flowtype for cb

File: src/networkLayer/client.js [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        return this._send(id, 'mutation', mutationRequest);
+    }
+
+    // We store the XHR as a RequestObject so that it can be retrieved and manipulated later
+    _storeRequest = (request: XMLHttpRequest, id: string) => {
+        const removeSelf = () => {
+            this.env._requests.delete(id);
+        };
+
+        const endOfLifeEvents = ['error', 'abort', 'loadend', 'timeout'];
```
Comment (sdcooke) [inline src/networkLayer/client.js]: Is it not just `load`?

File: src/networkLayer/client.js (hunk 4 above)
Comment (sdcooke) [inline src/networkLayer/client.js]: To be more specific:
 - csrf
 - on403callback
 - onSpeedbarResponse
 - onJailedCallback
shouldn't be in relay-helpers

and onMutationStartCallback and onMutationProgressCallback don't appear to be used

File: src/networkLayer/client.js (hunk 7 above)
Comment (JonathanUsername) [inline src/networkLayer/client.js]: I thought `load` fired when all data had been successfully read, whereas `loadend` fires when the data has finished transferring, and it always fires after error abort or load, so it's like a safety catch-all.

https://developer.mozilla.org/en-US/docs/Web/Events/loadend
https://www.nczonline.net/blog/2012/05/22/working-with-files-in-javascript-part-3/

**Pull request merged**